---
title: "seurat"
author: "Shuo Wang, quoted from Bingfei Yu"
date: "2023-09-09"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R script to run standard Seurat process on single-cell RNA data

### Preparation

```{r}
# Load required packges:
library(dplyr)
library(Seurat)
library(patchwork)
```

```{r}
# Load feature-cell matrix
s01.data <- Read10X(data.dir = "/Users/shuo/Desktop/filtered_feature_bc_matrix")
```

```{r}
# Initialize the Seurat object with the raw (non-normalized data).
s01 <- CreateSeuratObject(counts = s01.data, project = "s01preT", min.cells = 3, min.features = 200)
s01
```

### Filtering out low-quality cells

QC parameters:\
-nFeature_RNA: How many genes are detected per cell. I typically use 200 as a minimum but check the distribution in your data. Cells with fewer than 200 genes detected typically don't have enough information to cluster on.\
-nCount_RNA: How many UMIs or transcripts detected per cell. I typically don't filter using this metric as it is linearly correlated with nFeature_RNA. Some people will use an upper threshold here to try and filter out doublets but typically is not robust and we describe alternative approaches below.\
-percent.mt: Percent mitochondrial transcripts. Dying cells will have more mitochondrial gene content and so this can be used to filter low quality cells. Check the distribution in your data but typically upper thresholds of 5-10% mitochondrial content make sense.

```{r}
s01[["percent.mt"]] <- PercentageFeatureSet(s01, pattern = "^MT-")
```

```{r}
# Visualize QC metrics as a violin plot
VlnPlot(s01, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

```{r}
# Filter
length(Cells(s01))
s01 <- subset(s01, subset = nFeature_RNA > 200 & percent.mt < 10)
length(Cells(s01))
```

### Integration

### Normalize

Use the defaults pretty much always.

```{r}
s01 <- NormalizeData(s01)
```

### Find variable genes

Not a lot to modify here. If you have residual batch/unwanted variation (like cell cycle) effects after regression you can remove these genes from the variable gene set. Sometimes I also observed clustering based on variable immune genes (like TCR/Ig transcripts) so these can be removed as well.

```{r}
s01 <- FindVariableFeatures(s01, selection.method = "vst", nfeatures = 2000)
```

### Scale & regress out unwanted variation

`vars.to.regress`: If you have batch effects/sources of unwanted variation like cell cycle effects you can regress them out here. Some other common variables to regress are nCount_RNA, percent.mt (if you observe clustering based on these features & not due to remaining low quality cells).

```{r}
s01 <- ScaleData(s01, features = rownames(s01))
```
Note, `features` default is variable features.

#### Regress out cell cycle effects

Source: <https://satijalab.org/seurat/articles/cell_cycle_vignette>.

```{r}
# A list of cell cycle markers, from Tirosh et al, 2015, is loaded with Seurat.
# We can segregate this list into markers of G2/M phase and markers of S phase
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
```

We assign scores in the CellCycleScoring() function, which stores S and G2/M scores in object meta data, along with the predicted classification of each cell in either G2M, S or G1 phase.\
CellCycleScoring() can also set the identity of the Seurat object to the cell-cycle phase by passing set.ident = TRUE (the original identities are stored as old.ident). [The `set.ident` effect will be reflected in plot labels]Â  Please note that Seurat does not use the discrete classifications (G2M/G1/S) in downstream cell cycle regression. Instead, it uses the quantitative scores for G2M and S phase.

```{r}
s01 <- CellCycleScoring(s01, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

# view cell cycle scores and phase assignments
head(s01[[]])
```

```{r}
# Visualize the distribution of cell cycle markers across
RidgePlot(s01, features = c("PCNA", "TOP2A", "MCM6", "MKI67"), ncol = 2)
```

```{r}
# Running a PCA on cell cycle genes
s01 <- RunPCA(s01, features = c(s.genes, g2m.genes))
DimPlot(s01)
```

We score single cells based on the scoring strategy described in Tirosh et al. 2016. See `?AddModuleScore()` in Seurat for more information, this function can be used to calculate supervised module scores for any gene list.

For each gene, Seurat models the relationship between gene expression and the S and G2M cell cycle scores. The scaled residuals of this model represent a 'corrected' expression matrix, that can be used downstream for dimensional reduction.

```{r}
s01 <- ScaleData(s01, vars.to.regress = c("S.Score", "G2M.Score"), features = rownames(s01))
```

```{r}
# Run a PCA on only cell cycle genes again
s01 <- RunPCA(s01, features = c(s.genes, g2m.genes))
DimPlot(s01)
```

The procedure above removes all signal associated with cell cycle. In some cases, we've found that this can negatively impact downstream analysis, particularly in differentiating processes (like murine hematopoiesis), where stem cells are quiescent and differentiated cells are proliferating (or vice versa). In this case, regressing out all cell cycle effects can blur the distinction between stem and progenitor cells as well.

As an alternative, we suggest regressing out the difference between the G2M and S phase scores. This means that signals separating non-cycling cells and cycling cells will be maintained, but differences in cell cycle phase among proliferating cells (which are often uninteresting), will be regressed out of the data

```{r}
s01$CC.Difference <- s01$S.Score - s01$G2M.Score
s01 <- ScaleData(s01, vars.to.regress = "CC.Difference", features = rownames(s01))
```

### PCA

`dims`: The argument in `FindNeighbors`. Pick how many dimensions to use after PCA. I typically use ElbowPlot to decide. Including more PCs is usually better than too few, so I would err past the elbow. Between 10-40 PCs is typically appropriate.

```{r}
s01 <- RunPCA(s01, features = VariableFeatures(object = s01), verbose = FALSE)
DimPlot(s01, reduction = "pca")
```

#### DimHeatmap

```{r}
DimHeatmap(s01, dims = 1:15, cells = 500, balanced = TRUE)
```

#### Elbow plot

```{r}
ElbowPlot(s01, ndims = 50, reduction = "pca")
```

### Cluster

`resolution`: Argument to the FindClusters function. This is a big one you can change. Higher resolution values will give more clusters, lower resolution values give fewer clusters. How many clusters you want depends on the biological variation in your data as well as how you want to analyze it (are broad cell types enough or do you want specific cell states?). Values between 0.4-1.2 are typically appropriate.

```{r}
s01 <- FindNeighbors(s01, dims = 1:30)
s01 <- FindClusters(s01, resolution = 0.5)
```

### UMAP

We suggest using the same PCs as input to the clustering analysis. (<https://satijalab.org/seurat/articles/pbmc3k_tutorial>)

```{r}
s01 <- RunUMAP(s01, dims = 1:30)
```

```{r}
DimPlot(s01, reduction = "umap")
```

### Differential expression analysis

```{r}
s01.markers <- FindAllMarkers(s01, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
```

#### DE table

```{r}
write.csv(s01.markers, "s01-markers.csv")
```

#### Violin Plot

```{r}
VlnPlot(s01, features = c("CD3E", "CD4", "CD8A", "FOXP3", "CCL5"))
```

#### Feature Plot

```{r}
FeaturePlot(s01, features = c("CD3E", "CD4", "CD8A", "FOXP3", "CCL5"))
```

#### Heatmap

```{r}
s01.markers %>%
    group_by(cluster) %>%
    top_n(n = 10, wt = avg_log2FC) -> top10
DoHeatmap(s01, features = top10$gene) + NoLegend()
```

### Annotate

```{r}
new.cluster.ids <- c("CD4 T", "Treg", "CD4 T", "unkown T", "CD8 T", "unkown cell")
names(new.cluster.ids) <- levels(s01)
s01 <- RenameIdents(s01, new.cluster.ids)
```

#### DimPlot

```{r}
DimPlot(s01, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()
```

### After check

#### Cell quality

```{r}
VlnPlot(s01, features = c("nCount_RNA", "nFeature_RNA", "percent.mt"))
```

#### Cell cycle effect

```{r}
VlnPlot(s01, features = c("S.Score", "G2M.Score"))
```

#### Remove obvious low-quality clusters

```{r}
s01 <- subset(s01, idents = c("CD4 T", "Treg", "unkown T", "CD8 T"))
```

#### Save RDS

```{r}
saveRDS(s01, file = "s01.rds")
```

Note: Typically no need to normalize and scale the subsetted data again once the whole dataset has been fully processed. **But DoubletFinder suggests to rerun pre-process.**

### Quickly rerun pre-process
```{r}
s01 <- NormalizeData(s01)
s01 <- FindVariableFeatures(s01, selection.method = "vst", nfeatures = 2000)
s01 <- ScaleData(s01, features = rownames(s01))
s01 <- RunPCA(s01, features = VariableFeatures(object = s01), verbose = FALSE)
```

```{r}
ElbowPlot(s01, ndims = 50, reduction = "pca")
```

```{r}
s01 <- FindNeighbors(s01, dims = 1:30)
s01 <- FindClusters(s01, resolution = 0.5)
```

```{r}
s01 <- RunUMAP(s01, dims = 1:30)
DimPlot(s01, reduction = "umap")
```

```{r}
s01.markers <- FindAllMarkers(s01, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
write.csv(s01.markers, "s01-markers.csv")
```

```{r}
VlnPlot(s01, features = c("CD3E", "CD4", "CD8A", "FOXP3", "CCL5"))
```

```{r}
FeaturePlot(s01, features = c("CD3E", "CD4", "CD8A", "FOXP3", "CCL5"))
```

```{r}
s01.markers %>%
    group_by(cluster) %>%
    top_n(n = 10, wt = avg_log2FC) -> top10
DoHeatmap(s01, features = top10$gene) + NoLegend()
```

```{r}
new.cluster.ids <- c("CD4 T", "Treg", "CD4 T", "unkown T", "CD8 T")
names(new.cluster.ids) <- levels(s01)
s01 <- RenameIdents(s01, new.cluster.ids)
DimPlot(s01, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()
```

#### Save RDS

```{r}
saveRDS(s01, file = "s01_subseted.rds")
```

